libraries/tickfs/SPEC.md

```diff
@@ -0,0 +1,494 @@
+    after being defraged in memory
+ 1. Erase the reserved region
+
+The above steps would ensure we don't loose data on a power loss, as long as
+we could ensure we could continue the operation after regaining power.
+
+There are a few problems with this approach though.
+
+In order to maintain "garbage collecting" state to restore after a power loss
+we would need some sore of super block. This would then break the requirement
+on wear leveling as whichever region that block is in will be written and
+erased more often the others.
+
+The other problem is the reserved region where we would move data to. This
+would cost space as an entire region would be dedicated to garbage collecting.
+The garbage collecting region would also have more erase and writes performed
+on it breaking the wear leveling requirement.
+
+### Somewhat high storage overhead
+
+The storage overhead is somewhat high for TickFS. This is mostly due to the
+two 64-bit hashes that are stored with every object.
+
+The 64-bit value is determined by the output of the standard Rust
+`core::hash::Hasher` trait.
+
+For the hash of the key a 64-bit value can be justified by the lack of
+collision avoidance in the implementation. If two keys have the same
+hash the second key will be dropped. In this case a 64-bit hash should
+hopefully make that occurrence very unlikely. Also by following the standard
+Rust `core::hash::Hasher` trait the user is free to implement any standard
+Hasher of their choosing. Some systems will even be able to offload the
+hash to hardware.
+
+As for the check-sum. It was seriously considered to change this to a
+32-bit or even 16-bit CRC. This would reduce the overhead for each object
+by 4 or 6 bytes. Although a CRC is relatively easy to implement, it can still
+be time consuming. For this reason lots of embedded hardware will include a
+CRC offload engine to offload the calculation. By implementing a CRC in this
+library I would remove the option of allowing a offload engine. As there
+doesn't seem to be a generic CRC Rust trait to take advantage of the decision
+was made to reuse the `core::hash::Hasher` trait for the checksum. This costs
+space, but hopefully reduces implementation burden and allows for more hardware
+offloading.
+
+### Memory usage
+
+At a minimum TickFS requires a `PageSize` amount of free memory to store the
+buffer obtained when reading flash. This could be reduced by instead performing
+multiple reads of a smaller fixed size.
+
+### Synchronous
+
+TickFS is fully synchronous and every call is blocking.
+
+This is unfortunately inherently different to Tock's asynchronous design. The
+justification for this design is that in general writing and erasing flash
+can be tricky if we are executing from the same bank. A synchronous design
+seems like an overall more useful implementation to allow this. It also means
+we block in the critical moments where a power loss would loose the data we
+want to preserve.
```

Comment:
What does the complexity for executing from the same bank as the file system look like?

Reply:
That will be hardware dependent. Most flash controllers have limitations when writing to/from flash. For example you can't load other data from flash while performing an erase. In past experience usually interrupts are disabled during a flash erase to make it a critical section.

There is some discussion about this here: https://github.com/tock/tock/issues/1459#issuecomment-554742330

By making this a synchronous operation we can hide all of that complexity and just offload it to the `FlashController` implementation.

Reply:
Making an asynchronous operation look synchronous is challenging, but Tock has support calls for deferred calls which enable a synchronous operation to look asynchronous. If a flash controller driver needs to block during an erase it can, then it can set a deferred call to trigger a callback. However, if the flash controller knows the operation can happen asynchronously, it still has to block when it could instead wait for an interrupt.

Reply:
That's a good idea. The defered calls should let us re-use Tocks' Flash HIL with this current interface.